deepspeed/module_inject/containers/features/megatron.py

```diff
@@ -24,6 +24,22 @@ def transpose_qkv_alignment(self, x):
         else:
             return torch.cat((q.reshape(-1), k.reshape(-1), v.reshape(-1)), dim=-1).reshape(x.shape)
 
+    def _align_qkv(self, x):
+        attention_head_size = x.shape[0] // self.num_attention_heads
+        new_x_shape = (self.num_attention_heads, attention_head_size) + x.size()[1:]
+        x_1 = x.view(*new_x_shape)
+        div_dim = len(x_1.size()) - 1
+        (q, k, v) = torch.split(x_1, (x_1.shape[div_dim] // 3), dim=div_dim)
+        if len(q.shape) > 2:
+            return torch.cat((q.reshape(-1, q.shape[-1]), k.reshape(-1, q.shape[-1]), v.reshape(-1, q.shape[-1])),
+                             dim=0).reshape(x.shape)
+        else:
+            return torch.cat((q.reshape(-1), k.reshape(-1), v.reshape(-1)), dim=-1).reshape(x.shape)
+
+    def _update_merged_qkv(self):
+        self.qkvw = torch.nn.parameter.Parameter(self._align_qkv(self.qkvw).contiguous())
+        self.qkvb = torch.nn.parameter.Parameter(self._align_qkv(self.qkvb).contiguous())
```

Comment:
@RezaYazdaniAminabadi is self.qkvb only defined for containers inheriting from `features/megatron.py`? I wonder if we could isolate the `if self.qkvb is not None:` branches to the megatron container, rather than add to the base container?